xdsl/dialects/onnx.py

```diff
@@ -256,6 +256,98 @@ def verify_(self) -> None:
+    First input is the data tensor, second input is a shape tensor which specifies the output shape. It outputs the reshaped tensor.
+    At most one dimension of the new shape can be -1. In this case, the value is
+    inferred from the size of the tensor and the remaining dimensions. A dimension
+    could also be 0, in which case the actual dimension value is unchanged (i.e. taken
+    from the input tensor). Shape (second input) could be an empty shape, which means converting to a scalar.
+    The input tensor's shape and the output tensor's shape are required to have the same number of elements.
+    """
+
+    name = "onnx.Reshape"
+    T = Annotated[AnyFloat | IntegerType, ConstraintVar("T")]
+    data = operand_def(TensorType[T])
+    shape = operand_def(TensorType)
+    reshaped = result_def(TensorType[T])
+
+    allow_zero = opt_attr_def(IntegerAttr, attr_name="allowzero")
+
+    assembly_format = "`(` $data `,` $shape `)` attr-dict `:` `(` type($data) `,` type($shape) `)` `->` type($reshaped)"
+
+    def __init__(self, data: SSAValue, shape: SSAValue, allow_zero: Attribute):
+        super().__init__(
+            attributes={"allowzero": allow_zero},
+            operands=[data, shape],
+            result_types=[data.type],
+        )
+
+    def verify_(self) -> None:
+        if (
+            not isinstance(data_type := self.data.type, TensorType)
+            or not isinstance(shape_type := self.shape.type, TensorType)
+            or not isinstance(reshaped_type := self.reshaped.type, TensorType)
+        ):
+            assert (
+                False
+            ), "onnx elementwise operation operands and result must be of type TensorType"
+
+            data_type = cast(TensorType[Attribute], data_type)
+            reshaped_type = cast(TensorType[Attribute], reshaped_type)
+            shape_type = cast(TensorType[Attribute], shape_type)
+
+            if data_type != reshaped_type:
+                raise VerifyException("Mismatch between operand type and res type")
+            if shape_type.element_type != IntegerType(64):
+                raise VerifyException(
+                    "shape element type has to be a 64-bit signless integer"
+                )
+
+        data_type = data_type.get_shape()
+        shape_type = shape_type.get_shape()
+        shape_type_list = list(shape_type)
+        reshaped_type = reshaped_type.get_shape()
+
+        if (
+            self.allow_zero
+            and shape_type_list.count(0) > 0
+            and shape_type_list.count(-1) > 0
+        ):
+            raise VerifyException(
+                "Invalid shape: Both 0 and -1 are present, making the dimension ambiguous."
+            )
+        # At most one dimension of the new shape can be -1.
+        # In this case, the value is inferred from the size of the tensor and the remaining dimensions.
+        count_minus_one = shape_type_list.count(-1)
+        if count_minus_one == 1:
```

Comment:
what happens if there are more than `-1` specified dimension sizes? does mlir-onnx handle this (see the `ShapeHelper.cpp` if necessary)?

Reply:
seems onnx-mlir doesnt handle this, i looked at the helper and it seems it only supports shapes of rank one, so i dont think it has to be handled

Reply:
Yeah, I glanced at it again earlier and nothing obvious, I just find it weird that it is explicitly mentioned as a constraint in the documentation. I wonder what it does if given more than one.

Reply:
After looking at the docs, this refers to the contents of the 1-d shape tensor, so something like this:

```mlir
 %shape = onnx.Constant dense<[-1, -1, 2]> : tensor<3xi64>
```

would cause a runtime error, so there is nothing for this patch to act on.